JSONPath Support for TriggerBindings

dibyom opened the pull request:
# Changes
There are two commits in this PR. The first commit adds functions for JSONPath and the second one uses those functions to extract values for EventBinding.

Commits:
This commit adds functions to support JSONPath
in `TriggerBindings` using the `k8s.io/client-go/util/jsonpath`
library.

There are a few deviations from the default behavior of the library:

1. The JSONPath expressions have to be wrapped in the Tekton variable
interpolation syntax i.e `$()` : `$(body)`.

2.  We  support the `RelaxedJSONPathExpresion` syntax used in
`kubectl get -o custom-columns`. This means that the curly braces `{}`
and the leading dot `.` can be omitted i.e we support both `$(body.key)`
as well as `$({.body.key})

3. We return valid JSON values when the value is a JSON array or
map. By default, the library returns a string containing an internal
go representation. So, if the JSONPath expression selects `{"a": "b"}`,
the library will return `map[a: b]` while we return the valid JSON
string i.e `{"a":"b"}`

This commit switches TriggerBinding params to use
JSONPath instead of GJSON.

Part of #178 

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/triggers/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

```
TriggerBindings now support JSONPath expressions instead of GJSON. Check docs/triggerbindings.md for the syntax and examples

```

wlynch commented on pkg/template/jsonpath.go:
```diff
@@ -0,0 +1,155 @@
+// omit the leading curly braces '{}' and '.'
+func TektonJSONPathExpression(expr string) (string, error) {
+	if !isTektonExpr(expr) {
+		return "", errors.New("Expression not wrapped in $()")
```
Won't this error for cases of simple strings? We'll still want this for environment parameters. (not sure if we want to add this logic here, or above in ParseJSONPath)

dibyom replied:
Great catch...why didn't the test fail :thinking:

dibyom replied:
Added a check in the calling function. Need to add tests.

dibyom replied:
added tests in Test_NewResources

wlynch commented on pkg/template/event.go:
```diff
@@ -169,26 +169,65 @@ func getHeaderValue(header map[string][]string, headerName string) (string, erro
+
+// ApplyEventValuesToParams returns a slice of Params with the JSONPath variables replaced
+// with values from the event body and headers.
+func ApplyEventValuesToParams(ec Event, params []pipelinev1.Param) ([]pipelinev1.Param, error) {
```
super nit: my preference is for this to be (params []pipelinev1.Param, ec Event), since the params are the primary object we are acting on here.

dibyom replied:
done!

wlynch commented on pkg/template/jsonpath.go:
```diff
@@ -0,0 +1,155 @@
+	jsonRegexp = regexp.MustCompile(`^\{\.?([^{}]+)\}$|^\.?([^{}]+)$`)
+)
+
+func ParseJSONPath(input interface{}, expr string) (string, error) {
```
Add godoc comments to describe the inputs/outputs

dibyom replied:
done

wlynch commented on pkg/template/jsonpath.go:
```diff
@@ -0,0 +1,155 @@
+
+	// Copy of j.PrintResults
+	for ix := range fullResults {
+		if err := printResults(buf, fullResults[ix]); err != nil {
```
Can you provide guidance on how to interpret fullResults (`[][]reflect.Value`) vs results (`[]reflect.Value`)? Is this a distinction of line vs field?

dibyom replied:
so I think the only time the length of `[][]reflect.Value` is greater than 1 is when we use `range` keyword. AFAIK the `range` keyword is not standard JSONPath, its something that is only available in this library and used in kubectl

dibyom replied:
https://gist.github.com/so0k/42313dbb3b547a0f51a547bb968696ba#jsonpath

wlynch commented on pkg/template/jsonpath.go:
```diff
@@ -0,0 +1,155 @@
+		if i != len(results)-1 {
+			text = append(text, ' ')
+		}
+		if _, err = wr.Write(text); err != nil {
```
nit: `if _, err := ...`

wlynch commented on pkg/template/jsonpath.go:
```diff
@@ -0,0 +1,155 @@
+		if err != nil {
+			return nil, err
+		}
+		return text, nil
```
`return json.Marshal(v.Interface())`

dibyom replied:
done

wlynch commented on pkg/template/jsonpath.go:
```diff
@@ -0,0 +1,155 @@
+		if i != len(results)-1 {
+			text = append(text, ' ')
+		}
+		if _, err = wr.Write(text); err != nil {
```
Not sure if this is worth doing or not, but thoughts on doing things this way as opposed to passing the writer through?

dibyom replied:
What's the advantage? My thinking was to keep the copied code from the library (i.e. the printResults and the evalToText function) as close to the original as possible, try to upstream the fix in  the original library itself, and then at some point in the future remove the copy pasta!

wlynch replied:
SGTM

wlynch commented on pkg/template/jsonpath.go:
```diff
@@ -0,0 +1,155 @@
+	if len(pathExpression) == 0 {
+		return pathExpression, nil
+	}
+	submatches := jsonRegexp.FindStringSubmatch(pathExpression)
```
Mostly because I don't fully grok the regex being used here, but what exactly are we trying to extract here? (might be helpful to provide an example).

dibyom replied:
The function docstring and the substring error has some examples of what the regex accepts but basically it tries to extract a substring - so:
`{blah}` or `{.blah}` or `.blah` or `blah`

The substring however cannot contain curly braces so `bla{h` won't match

wlynch commented on pkg/template/jsonpath.go:
```diff
@@ -0,0 +1,155 @@
+	return fmt.Sprintf("{.%s}", fieldSpec), nil
+}
+
+// unwrwapTektonExpr takes an experession of the form $(abc) and returns abc.
```
unwrapTektonExpr, expression

dibyom replied:
🤕

wlynch commented on pkg/template/jsonpath_test.go:
```diff
@@ -0,0 +1,144 @@
+	tests := []struct {
+		name  string
+		expr string
+		in     string
```
Instead of encoding the object body to a string then unmarshalling it out, would it be easier to make this interface{} and set the fields as maps?

dibyom replied:
yeah, don't have a strong opinion on this -> might be a bit more verbose for the array and objects that contain other embedded objects?

wlynch commented on pkg/template/jsonpath_test.go:
```diff
@@ -0,0 +1,144 @@
+		name: "null values",
+		in: objectBody,
+		expr: "$(body.null)",
+		want: "null",
```
Is this WAI? Even though the payload says `null` we should probably make a distinction between `null` and `"null"`. I say either default to `""` or have ParseJSONPath return `*string` so we can represent this nil value (my guess is `""` is probably sufficient)

dibyom replied:
Yeah, so I went back and forth on this a few times before settling on "null". I agree that if we were talking about return go values, the correct way should be `*string`. However, for our use case here (i.e. concatenating the json values in string in the trigger templates), that would result in the string `<nil>`.  If we use "" to represent the printable value of `null`, we have no way to differentiate between `null` and the empty string "".

The `"null"` and `null` collision is unfortunate though I think this is the least bad way at the moment. Also, this issue is only valid if the JSONPath selects a field whose value is `null`. If the selection is an object or an array which contains both the string null as well as a null value, we get the following: `{"stringNull":"null","nullValue":null}`

dibyom commented on pkg/template/event.go:
```diff
@@ -169,26 +169,65 @@ func getHeaderValue(header map[string][]string, headerName string) (string, erro
+}
+
+// NewEvent returns a new Event from HTTP headers and body
+func NewEvent(body []byte, headers map[string][]string) (Event, error) {
```
done

dibyom commented on pkg/template/event.go:
```diff
@@ -169,26 +169,65 @@ func getHeaderValue(header map[string][]string, headerName string) (string, erro
+	}
+	joinedHeaders := make(map[string]string, len(headers))
+	for k, v := range headers {
+		joinedHeaders[k] = strings.Join(v, ",")
```
Ack. While we do use ArrayOrString in Param, we only ever use the string value, so I think its simpler to just use strings as much as possible expect when we have to assign the value to the Param.

dibyom commented on pkg/template/event.go:
```diff
@@ -169,26 +169,65 @@ func getHeaderValue(header map[string][]string, headerName string) (string, erro
+
+
+// Event represents a HTTP event that Triggers processes
+type Event struct {
```
Not really at the moment. Though it would be nice if Filters could reuse the same struct later...unexported for now

dibyom commented on pkg/template/jsonpath.go:
```diff
@@ -0,0 +1,155 @@
+		return "", err
+	}
+
+	// Copy of j.PrintResults
```
Old comment...don't think its really applicable anymore

dibyom commented on pkg/template/jsonpath.go:
```diff
@@ -0,0 +1,155 @@
+
+	// Copy of j.PrintResults
+	for ix := range fullResults {
+		if err := printResults(buf, fullResults[ix]); err != nil {
```
There isn't I think besides following the pattern from the original library at https://github.com/tektoncd/triggers/blob/master/vendor/k8s.io/client-go/util/jsonpath/jsonpath.go#L72

Changed!

dibyom commented on pkg/template/event.go:
```diff
@@ -169,26 +169,65 @@ func getHeaderValue(header map[string][]string, headerName string) (string, erro
-		return nil, xerrors.Errorf("error merging trigger params: %v", err)
+		return nil, xerrors.Errorf("error merging trigger params: %q", err)
 	}
-	out, err = ApplyBodyToParams(body, out)
```
Only used in tests. Wanted to remove it in a different PR to keep the size of this small but can also do it here.

wlynch replied:
I think it's fine to remove it here. Simple deletions are easy to grok, and I'd rather have a larger PR than someone accidentally reading / trying to modify a dead piece of code. Also helps in the event of a rollback (only 1 commit to revert instead of 2).

dibyom commented on pkg/template/event.go:
```diff
@@ -169,26 +169,65 @@ func getHeaderValue(header map[string][]string, headerName string) (string, erro
+}
+
+// NewEvent returns a new Event from HTTP headers and body
+func NewEvent(body []byte, headers map[string][]string) (Event, error) {
```
done

dibyom commented on pkg/template/jsonpath_test.go:
```diff
@@ -0,0 +1,145 @@
+		name: "objects",
+		in:   objectBody,
+		expr: "$(body)",
+		// TODO: Do we need to escape backslashes for backwards compat?
```
@ncskier @vtereso In the current gjson bindings, we always escape the quotes `\"` in the returned value. Why do we do that?

ncskier replied:
I think it's because we're passing we're passing that JSON object as a type String in the ResourceTemplate JSON: https://github.com/tektoncd/triggers/blob/master/pkg/sink/sink.go#L195

dibyom replied:
ok, I'm adding escaping behavior for now for backwards compatibility. I'll open an issue to explore alternative approaches here -- maybe smarter concatenation somehow or using json patches

ncskier replied:
Sounds good, thanks!

dibyom commented:
/retest

wlynch commented on pkg/template/jsonpath.go:
```diff
@@ -0,0 +1,158 @@
+
+	// jsonRegexp is a regular expression for JSONPath expressions
+	// with or without the enclosing {} and the leading . inside the curly
+	// braces e.g.  'a.b' or '.a.b' or '{a.b}' or '{.a.b}
```
nit: '{.a.b}'

dibyom replied:
doh

dibyom commented:
> LGTM pending discussion of #253, since IIUC this will break bindings that mix static and interpolated values.

Yeah, let's keep backwards compatibility for now. If we decide to go with #253 later it should be simple enough to undo.

bobcatfish commented on pkg/template/jsonpath_test.go:
```diff
@@ -0,0 +1,145 @@
+			got, err := template.ParseJSONPath(data, tt.expr)
+			if err != nil {
+				t.Errorf("ParseJSONPath() error = %v", err)
+				return
```
although using `Fatalf` in a goroutine is dangerous, it seems like it's fine to use `Fatalf` in a sub test, the golang blog even does it (https://blog.golang.org/subtests) so i think we you could use Fatalf here instead of Errorf + return

bobcatfish replied:
(same for other occurrences of this)

dibyom replied:
Neat!

bobcatfish replied:
i think this line could be fatalf and then no return needed

bobcatfish commented on pkg/template/jsonpath_test.go:
```diff
@@ -0,0 +1,145 @@
+				return
+			}
+
+			if diff := cmp.Diff(got, strings.Replace(tt.want, " ", "", -1)); diff != "" {
```
i think `cmp.Diff` output is way easier to read if you pass in "want, got" (e.g. like in this example https://godoc.org/github.com/google/go-cmp/cmp#Diff), then you intuitively get a - besides things you wanted but didnt get, and a + for things you didnt want that you got (the other way around is confusing imo)

dibyom replied:
Flipped!

bobcatfish commented on pkg/template/jsonpath.go:
```diff
@@ -0,0 +1,158 @@
+package template
```
hm do you think there's another name we could use besides "template"? in pipelines we are trying to avoid using the term "templating" for the variable substitution we are doing b/c "template" implies additional functionality we want to avoid adding

bobcatfish replied:
i just realized the type is literally called "trigger template" and this package is probably named after that so just ignore me :woman_facepalming:

dibyom replied:
Given the discussion around JSONPath support in pipeline params, it might be worth having a common jsonpath pkg that both pipeline and triggers could use :thinking: 

(Though that's a yak that we should shave another day :smile:

bobcatfish commented on pkg/template/jsonpath_test.go:
```diff
@@ -0,0 +1,145 @@
+	tests := []struct {
+		expr    string
+		want    string
+		wantErr bool
```
i like how the above tests are separate for the error case and the success case, could we do that here as well?

dibyom replied:
done!

bobcatfish commented on pkg/template/resource.go:
```diff
@@ -92,7 +93,10 @@ func ApplyParamsToResourceTemplate(params []pipelinev1.Param, rt json.RawMessage
 	paramVariable := fmt.Sprintf("$(params.%s)", param.Name)
-	return bytes.Replace(rt, []byte(paramVariable), []byte(param.Value.StringVal), -1)
+	// Escape quotes for backwards compatibility. Needed since we are appending JSON
+	//
```
extra comment line?

also is this temporary, if so can we have an issue to track removing it? and if it isn't temporary then maybe "backward compatibility" is a bit vague and we can explain a bit more about what we are being backward compatible with and why?

dibyom replied:
Opened #257 to discuss this.

bobcatfish commented on pkg/template/jsonpath.go:
```diff
@@ -0,0 +1,158 @@
+
+// unwrapTektonExpr takes an expression of the form $(abc) and returns abc.
+func unwrapTektonExpr(expr string) string {
+	return tektonVar.ReplaceAllString(expr, "$1")
```
im not a huge fan of functions that only have one line in them, i feel like they just add a layer of indirection without adding too much (e.g. if i want to know what `unwrapTektonExpr` does i have to come over here and look at this vs looking direclty at `tektonVar.ReplaceAllString`), but maybe there is a benefit i'm missing?

dibyom replied:
removed

bobcatfish commented on pkg/template/jsonpath.go:
```diff
@@ -0,0 +1,158 @@
+// If the value if an array or map, it returns a JSON representation
+// of the value (as opposed to the internal go representation of the value)
+// Otherwise, the text value is from the `evalToText` function, originally from
+// k8s.io/client-go/util/jsonpath/jsonpath.go
```
it seems like there are a few functions in this file that are copied directly from other places - maybe your commit message could explain a bit more about why? (im thinking it's b/c they were unexported and you need to use them?)

dibyom replied:
added to commit message!

dibyom commented on pkg/template/event.go:
```diff
@@ -18,177 +18,79 @@ package template
 	out, err := MergeBindingParams(bindings)
 	if err != nil {
-		return nil, xerrors.Errorf("error merging trigger params: %v", err)
+		return nil, xerrors.Errorf("error merging trigger params: %q", err)
```
1. What about %w which calls `error.UnWrap` for errors and %v for everything else?
2. TIL!

dibyom replied:
Done!

bobcatfish replied:
ive never used `%w` before! TIL :D

dibyom commented on pkg/template/event_test.go:
```diff
@@ -28,628 +27,7 @@ import (
-	}
-}
-
+// TODO: 2. Split NewResourcesTests into separate tests for ResolveParams and ResolveResources
```
Ahh, I opened the issue and forgot to update the comment :man_facepalming:

dibyom commented on pkg/template/event.go:
```diff
@@ -18,177 +18,79 @@ package template
+// with values from the event body and headers.
+func applyEventValuesToParams(params []pipelinev1.Param, ec *event) ([]pipelinev1.Param, error) {
+	for idx, p := range params {
+		// TODO: We need tests for binding values that are not jsonpath expressions
```
Actually added the tests so its not a TODO anymore :sweat_smile:

dibyom commented on docs/triggerbindings.md:
```diff
@@ -33,46 +33,44 @@ Each parameter has a `name` and a `value`.
-`\$\(header(\.[[:alnum:]_\-]+)?\)`
+TriggerBindings can access values from the HTTP JSON body and the headers using
+JSONPath expressions. The expressions have to be wrapped in `$()` but can omit
+the curly braces `{}` and the leading `.`. 
```
https://kubernetes.io/docs/reference/kubectl/jsonpath/ mentions the enclosing `{}` though it also says its a JSONPath "template"....maybe we keep things simple and never mention the `{}`?

dibyom replied:
Removed the curly braces from the examples!

dibyom commented on pkg/template/event.go:
```diff
@@ -18,177 +18,79 @@ package template
 }
+
+// event represents a HTTP event that Triggers processes
+type event struct {
```
might be confusing with `http.Request`. Also, I like event since we can later add the one-offs such as the `$(uid)` to this struct to unify how we do the "templating"

bobcatfish replied:
> might be confusing with http.Request

it basically _is_ an http.Request tho since it has a header + a body :thinking: 

I see what you mean about the `uid` tho! you're probably right about `event` being better in the long run, sgtm :D

dibyom commented on pkg/template/event.go:
```diff
@@ -18,177 +18,79 @@ package template
+
+	return &event{
+		Header: joinedHeaders,
+		Body:   data,
```
Yeah I added that check after verifying the empty body case: https://github.com/tektoncd/triggers/pull/241/files/4fe79f8e2d17b38e31d2d871c983a545cb06db01#diff-8d7c043b29f6a04a92258e1aac64fc2bR44

bobcatfish replied:
i cant see that link anymore - good ol rebasing :sweat_smile:

dibyom replied:
:face_with_head_bandage:

dibyom replied:
https://github.com/tektoncd/triggers/pull/241/files?file-filters%5B%5D=.go#diff-8d7c043b29f6a04a92258e1aac64fc2bR42

dibyom commented:
> i think mostly you've done a great job, e.g. unit tests for ParseJSONPath but there are couple of functions which i think could also use tests which you probably skipped b/c they are unexported, and i think many of the functions (e.g. relaxedJSONPathExpression) are significant enough to actually be exported and have their own tests

The current `TektonJSONPathExpression` tests does test all the behaviors of RelaxedJSONPath expression. I do see the point about the behavior being easier to grok if some of the unexported functions have their own tests though I'm not sure if we should be exporting functions solely because they are complex or for testing.

dibyom commented:
/test pull-tekton-triggers-build-tests

dibyom commented:
Ok, I think I addressed all the comments! 

Also, added a few more tests (header values, more relaxed jsonpath expressions etc.) 

There area a couple of if branches that are hard to test e.g. we only reach there if the unmarshalled value is a Function or a Channel --I don't think any HTTP JSON payload can be unmarshalled to that!
 
Quite a few followups from this:
* Escaping strings (#258, #257)
* Values for headers are sometimes inconsistent (#260)
* "Typed" Binding values? (#253)
* Keep existing behavior for 0.2, will look into changing for 0.3/
* More focussed tests (#252)

/hold cancel

bobcatfish commented:
Thanks for all the follow up issues @dibyom !! :heart:

bobcatfish commented on pkg/template/jsonpath_test.go:
```diff
@@ -0,0 +1,201 @@
+	err := json.Unmarshal([]byte(testJSON), &data)
+	if err != nil {
+		t.Fatalf("Could not unmarshall body : %q", err)
+		return
```
if you fatalf then you wont need to return b/c fatalf will halt execution

dibyom replied:
done!

bobcatfish commented:
/lgtm

bobcatfish commented:
/meow space
